### PR TITLE
no-extra-arguments: provide secondary issue locations

### LINKED ISSUE #95: no-extra-arguments: provide secondary issue locations

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #95

### CODE REVIEW

File: src/utils/locations.ts [hunk 1]
```diff
@@ -20,6 +20,20 @@
 import { Rule } from "eslint";
 import * as estree from "estree";
 
+export interface IssueLocation {
+  column: number;
+  line: number;
+  endColumn?: number;
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: remove optionality

File: tests/rules/no-extra-arguments.test.ts [hunk 2]
```diff
@@ -172,3 +195,7 @@ function message(
     ...extra,
   };
 }
+
+function checkEncodedMessage(expected: number, provided: number, secondaryLocations: IssueLocation[]) {
```
Comment (vilchik-elena) [inline tests/rules/no-extra-arguments.test.ts]: `encodedMessage` (to be consistent with `message` above)

File: tests/rules/no-extra-arguments.test.ts [hunk 3]
```diff
@@ -96,6 +107,18 @@ ruleTester.run("no-extra-arguments", rule, {
       `,
       errors: [message(0, 1, { line: 5, column: 9, endColumn: 15 })],
     },
+    {
+      code: `
+        var foo = function() {
```
Comment (vilchik-elena) [inline tests/rules/no-extra-arguments.test.ts]: add comment with location

File: src/utils/locations.ts [hunk 4]
```diff
@@ -61,6 +75,40 @@ export function getMainFunctionTokenLocation(
...
+ * Encode those extra information in the issue message when rule is executed
+ * in Sonar* environment.
+ */
+export function reportWithSecondaryLocationsAndCost(
+  context: Rule.RuleContext,
+  { secondaryLocations, cost, message, ...descriptor }: Rule.ReportDescriptor & EncodedMessage,
+) {
+  if (context.options[context.options.length - 1] === "sonar-runtime") {
+    const encodedMessage: EncodedMessage = { secondaryLocations, message };
+    if (cost) {
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: drop condition

File: src/utils/locations.ts [hunk 5]
```diff
@@ -61,6 +75,40 @@ export function getMainFunctionTokenLocation(
...
 }
 
+/**
+ * Wrapper for `context.report`, supporting secondary locations and cost.
+ * Encode those extra information in the issue message when rule is executed
+ * in Sonar* environment.
+ */
+export function reportWithSecondaryLocationsAndCost(
+  context: Rule.RuleContext,
+  { secondaryLocations, cost, message, ...descriptor }: Rule.ReportDescriptor & EncodedMessage,
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: move in the function

File: src/utils/locations.ts [hunk 6]
```diff
@@ -61,6 +75,40 @@ export function getMainFunctionTokenLocation(
   return location!;
 }
 
+/**
+ * Wrapper for `context.report`, supporting secondary locations and cost.
+ * Encode those extra information in the issue message when rule is executed
+ * in Sonar* environment.
+ */
+export function reportWithSecondaryLocationsAndCost(
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: let's call it `report(context, cost?, secondaries?, reportDescriptor)`

File: src/utils/locations.ts [hunk 7]
```diff
@@ -61,6 +75,40 @@ export function getMainFunctionTokenLocation(
...
+  context: Rule.RuleContext,
+  { secondaryLocations, cost, message, ...descriptor }: Rule.ReportDescriptor & EncodedMessage,
+) {
+  if (context.options[context.options.length - 1] === "sonar-runtime") {
+    const encodedMessage: EncodedMessage = { secondaryLocations, message };
```
Comment (vilchik-elena) [inline src/utils/locations.ts]:
```
if (has property){
const encodedMessage = { secondaryLocations, message, cost };
descriptor.message= JSON.string(encodedMessage);
}

context.report(descriptor);
```

File: src/utils/locations.ts [hunk 8]
```diff
@@ -61,6 +75,46 @@ export function getMainFunctionTokenLocation(
   return location!;
 }
 
+// As `ReportDescriptor` may contain either `message` or `messageId` prop,
+// we force the presence of `message` property by using the following type alias.
+export type ReportDescriptor = Rule.ReportDescriptor & { message: string };
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: Why are we exporting this type?

File: src/utils/locations.ts [hunk 9]
```diff
@@ -61,6 +75,46 @@ export function getMainFunctionTokenLocation(
...
+    const encodedMessage: EncodedMessage = { secondaryLocations, message, cost };
+    reportDescriptor.message = JSON.stringify(encodedMessage);
+  }
+  context.report(reportDescriptor);
+}
+
+/**
+ * Converts `SourceLocation` range into `IssueLocation`
+ */
+export function issueLocation(
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: I think it will be easier if this function takes nodes

File: tests/rules/no-extra-arguments.test.ts [hunk 10]
```diff
@@ -87,6 +88,18 @@ ruleTester.run("no-extra-arguments", rule, {
         message(2, 4, { line: 4, column: 9, endColumn: 24 }),
       ],
     },
+    {
+      code: `
+        function foo(p1, p2) {}
+        //           ^^^^^^<
```
Comment (vilchik-elena) [inline tests/rules/no-extra-arguments.test.ts]: arrow should be in opposite direction

File: src/utils/locations.ts (hunk 9 above)
Comment (andrea-guarino-sonarsource) [reply src/utils/locations.ts]: As discussed, we can keep it with SourceLocation

File: src/utils/locations.ts (hunk 8 above)
Comment (andrea-guarino-sonarsource) [reply src/utils/locations.ts]: Because it's used in an exported API as parameter

**Pull request merged**